akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala

```diff
@@ -128,12 +187,69 @@ private[akka] object Running {
       applyEffects(cmd, state, effect.asInstanceOf[EffectImpl[E, S]]) // TODO can we avoid the cast?
     }
 
+    def onReplicatedEvent(
+        state: Running.RunningState[S],
+        envelope: ReplicatedEventEnvelope[E]): Behavior[InternalProtocol] = {
+      // FIXME set the details on the context https://github.com/akka/akka/issues/29258
+      setup.log.info(
+        "Replica {} received replicated event {}. Replicas seq nrs: {}",
+        setup.activeActive,
+        envelope,
+        state.seenPerReplica)
+      envelope.ack ! ReplicatedEventAck
+      if (envelope.event.originReplica != setup.activeActive.get.replicaId && !alreadySeen(envelope.event)) {
+        setup.log.info("Saving event as first time")
+        handleReplicatedEventPersist(envelope.event)
+      } else {
+        setup.log.info("Filtering event as already seen")
+        this
+      }
+    }
+
     // Used by EventSourcedBehaviorTestKit to retrieve the state.
     def onGetState(get: GetState[S]): Behavior[InternalProtocol] = {
       get.replyTo ! state.state
       this
     }
 
+    private def handleReplicatedEventPersist(event: ReplicatedEvent[E]): Behavior[InternalProtocol] = {
+      _currentSequenceNumber = state.seqNr + 1
+      val newState: RunningState[S] = state.applyEvent(setup, event.event)
+      val newState2: RunningState[S] = internalPersist(setup.context, null, newState, event, "")
+      setup.log.info("State after apply {} state after internal persist {}", newState.state, newState2.state)
+      val shouldSnapshotAfterPersist = setup.shouldSnapshot(newState2.state, event.event, newState2.seqNr)
+      // FIXME validate this is the correct sequence nr from that replica https://github.com/akka/akka/issues/29259
+      val updatedSeen = newState2.seenPerReplica.updated(event.originReplica, event.originSequenceNr)
+      persistingEvents(
+        newState2.copy(seenPerReplica = updatedSeen),
+        state,
+        numberOfEvents = 1,
+        shouldSnapshotAfterPersist,
+        Nil)
+    }
+
+    private def handleEventPersist(event: E, cmd: Any, sideEffects: immutable.Seq[SideEffect[S]]) = {
+      // apply the event before persist so that validation exception is handled before persisting
+      // the invalid event, in case such validation is implemented in the event handler.
+      // also, ensure that there is an event handler for each single event
```

Comment:
With active active I guess applying the event could be more likely to fail since when the event is replicated another replica did the validation of a command and accepted it.

Reply:
This can also happen during the rolling deploy if an entity is moved back to an old node. I'll create a ticket for how we best handle the rolling update scenarios

Reply:
I think you'll have to be really careful and do two-step deploys, one introducing new events and then when that has rolled out you actually add the command handler logic that creates them. We need to spell that out in docs.